### DISCUSSION TITLE
Change names for shallow nested routes

### DISCUSSION THREAD

**Discussion by carestad** [laravel/framework#45882] [Q&A]
I'm looking to try and control how the shallow routes are registered. As an example I have the following route registered:

```php
Route::apiResource('articles', ArticleController::class);
Route::apiResource('articles.comments', ArticleCommentController::class)->shallow();
```

Which in turn registers the following routes:

- `GET|HEAD /articles`
- `POST /articles`
- `GET|HEAD /articles/{article}`
- `PUT|PATCH /articles/{article}`
- `DELETE /articles/{article}`
- `GET|HEAD /articles/{article}/comments`
- `POST /articles/{article}/comments`
-  `GET|HEAD /comments/{comment}`
- `PUT|PATCH /comments/{comment}`
- `DELETE /comments/{comment}`

I am wondering if I can somehow make it prefix the `/comments` routes with `article-` so they become:

-  `GET|HEAD /article-comments/{comment}`
- `PUT|PATCH /article-comments/{comment}`
- `DELETE /article-comments/{comment}`

Does anyone know if this can be done somehow? I couldn't find anything regarding this in the documentation or the discussion page here (or Google for that matter).

**Comment by denizgolbas** [comment]
```php
Route::apiResource('articles', ArticleController::class);
Route::apiResource('article-comments', ArticleCommentController::class, ['parameters' => [
            'article-comments' => 'comment',
        ]])->shallow(); //article-comments/{comment}
```

**Reply by carestad** [reply]
Thanks for the reply.

The parameters option will only rename the auto generated parameter names in the routes though, not the actual prefix. But I see the example you provided registered it as `article-comments` instead of `articles.comments` too, which will of course do this, but at that point `->shallow()` is also unnecessary.

It sort of does what I was looking for, although I was hoping to be able to keep the nested syntax (`articles.comments`) and the GET and POST routes for comments that are bound to the `/articles/{article}/comments`.

**Comment by benotoho** [comment]
Is there anyone to tag? I have the same problem as @carestad.

**Comment by S-Nidhin** [comment]
Yes, you can make use of route prefixing in Laravel to achieve the desired URL structure. In your case, you can prefix the /comments routes with article-, so they become /article/{article}/comments instead of /comments/{comment}.

To accomplish this, you can define a route group in your routes/web.php file or any other appropriate route file. Here's an example of how you can achieve the route prefixing:

Route::prefix('article/{article}')->group(function () {
    // Shallow routes for comments
    Route::get('/comments', [CommentController::class, 'index'])->name('article.comments.index');
    Route::post('/comments', [CommentController::class, 'store'])->name('article.comments.store');
    Route::get('/comments/{comment}', [CommentController::class, 'show'])->name('article.comments.show');
    Route::put('/comments/{comment}', [CommentController::class, 'update'])->name('article.comments.update');
    Route::delete('/comments/{comment}', [CommentController::class, 'destroy'])->name('article.comments.destroy');
});
In the above example, the route group is prefixed with 'article/{article}', which means all the routes inside this group will have the prefix /article/{article}. The {article} parameter will be available in the route's controller methods.

Make sure to replace CommentController with the actual name of your controller handling the comments.

By using this approach, you will achieve the desired URL structure where the /comments routes are prefixed with /article/{article}.As an experienced Laravel developer working in a prominent[ Laravel consulting ](https://cubettech.com/services/laravel-development-service/)company, I can confidently say that the modified shallow nested routes you provided appear to be a viable solution. By incorporating the route prefix and custom names, you have effectively organized the routes for comments under the /article/{article} structure. This approach enhances readability and maintainability while ensuring a clear hierarchical relationship between articles and their associated comments.

**Reply by benotoho** [reply]
Thank you for your response, but the entire point of the `shallow()` method is presumably to avoid requiring the parent ID (in this case, article ID) when the child ID (comment ID) is already a unique identifier within the application. `GET|HEAD /article-comments/{comment}`, for instance, is not achievable by following your example.

Sure we can always define a route for each HTTP verb separately, but the point of the `shallow()` method is convenience. This is how I was able to implement the 'article' prefix in my case:

```php
Route::resource('articles.comments', ArticleCommentController::class)->only(['index', 'store']);
Route::prefix('articles')->group(function() {
    Route::resource('/comments', ArticleCommentController::class)->only(['show', 'update', 'destroy']);
});
```
